[red-knot] Disjointness for callable types (#17094)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/type_properties/is_disjoint_from.md

static_assert(not is_disjoint_from(tuple[Literal[1], Literal[2]], tuple[Literal[1], int]))
```

> If a tuple type contains a `Never` element, it is considered similar to `Never` itself and is
> disjoint from any other type.

Reviewer: Maybe expand on _why_ it's similar to `Never`:

```suggestion
If a tuple type contains a `Never` element, then it cannot be instantiated. Like `Never`, the
type contains no objects, and is therefore disjoint from any other type.
```

> contains no objects, and is therefore disjoint from any other type.

Reviewer 2: It's great to add some tests for this. But I think I'd prefer just some tests like this; I think they'd be clearer:

```py
def f(x: tuple[Never], y: tuple[int, Never], z: tuple[Never, int]):
    reveal_type(x)  # revealed: Never
    reveal_type(y)  # revealed: Never
    reveal_type(z)  # revealed: Never
```

> contains no objects, and is therefore disjoint from any other type.

Reviewer: > I think this comment is slightly confusing. `tuple[int, Never]` is not _like_ `Never` in our model -- it literally _is_ `Never`!

Ah, got it! Does that suggest that we want this to be an explicit invariant of our representation? Is our intent to have `Type::Never` be the _only_ empty type? Are there other types that we would need to make sure simplify to `Never`? (I know intersection with `Never` is already simplified like this.)

> contains no objects, and is therefore disjoint from any other type.

Reviewer 2: It would certainly be nice if we _can_ maintain this invariant going forward! It makes a lot of things simpler to think about 😄 I can't say for sure whether it will always be possible to eagerly detect (and simplify to `Never`) all types that are known to be empty, though

> contains no objects, and is therefore disjoint from any other type.

Reviewer 3: Yes, I think we want this to be an invariant of our representation -- and I think it should be possible.

And I also agree that it would be better to test this simplification not in this file, and not using `is_disjoint_from`, but rather the way Alex suggests, and over in the tuple type tests.

> contains no objects, and is therefore disjoint from any other type.

Author: > Yes, I think we want this to be an invariant of our representation -- and I think it should be possible.

I can try and create a follow-up for this.

> And I also agree that it would be better to test this simplification not in this file, and not using `is_disjoint_from`, but rather the way Alex suggests, and over in the tuple type tests.

Make sense

```py
static_assert(is_disjoint_from(tuple[Never], tuple[Never]))
static_assert(is_disjoint_from(tuple[int, Never], tuple[int, Never]))
static_assert(is_disjoint_from(tuple[int, Never], tuple[Never, int]))
static_assert(is_disjoint_from(tuple[int, Never], tuple[int, int]))
```

> 

Author: This is not related to this PR but it helped my understanding when I discussed `Never` usage with Alex and David last week so I added some test cases for it as they were missing.

static_assert(is_disjoint_from(type[UsesMeta1], type[UsesMeta2]))
```

## Callables

No two callable types are disjoint because there exists a callable type

> `(*args: object, **kwargs: object) -> Never` that is a subtype of all fully static callable types.

Reviewer 2: (optional)

```suggestion
`(*args: object, **kwargs: object) -> Never` that is a subtype of all fully static callable types.
As such, for any two callable types, it is possible to conceive of a runtime callable object that would
inhabit both types simultaneously.
```

> No two callable types are disjoint because there exists a callable type
> `(*args: object, **kwargs: object) -> Never` that is a subtype of all fully static callable types.

Reviewer 3: Similar to Alex's suggestion, but I think it's important to clarify that it's a non-empty subtype, because even disjoint types have the common subtype `Never`. And personally I think once we clarify that, it's sufficient without adding an entirely new sentence: if two types have a non-empty common subtype, then obviously they share some inhabitants. But I don't feel strongly, we can add another sentence driving that point home.

```suggestion
No two callable types are disjoint because there exists a non-empty callable type
`(*args: object, **kwargs: object) -> Never` that is a subtype of all fully static callable types.
```
